### PR TITLE
HSEARCH-3105 Lucene type handling changes

### PR DESCRIPTION
**Pull request by gsmet**
* https://hibernate.atlassian.net/browse/HSEARCH-3105

I only did Lucene for now so that we can see how it feels.

I'm a bit worried about how many classes we will have in the end when we will implement all the types. I hope we'll be able to factorize them somehow.

Better reviewed commit per commit.

### CODE REVIEW

File: backend-lucene/src/main/java/org/hibernate/search/v6poc/backend/lucene/document/model/impl/LuceneFieldEncoding.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ */
+public interface LuceneFieldEncoding<T> {
+
+	void addFields(LuceneDocumentBuilder documentBuilder, LuceneIndexSchemaObjectNode parentNode, String fieldName, T value);
+
+	default Set<String> getOverriddenStoredFields() {
+		return Collections.emptySet();
+	}
+
+	T parse(Document document, String fieldName);
```
[the remark this answers is not in the document]
Comment (gsmet) [reply backend-lucene/src/main/java/org/hibernate/search/v6poc/backend/lucene/document/model/impl/LuceneFieldEncoding.java]: I changed the class to `Codec` and the methods to `encode()`/`decode()`.

File: backend-lucene/src/main/java/org/hibernate/search/v6poc/backend/lucene/types/impl/predicate/LocalDateFieldQueryFactory.java [hunk 2]
```diff
@@ -6,56 +6,27 @@
...
-import org.apache.lucene.search.Query;
 import org.hibernate.search.v6poc.backend.lucene.document.model.impl.LuceneFieldQueryFactory;
-import org.hibernate.search.v6poc.backend.lucene.document.model.impl.MatchQueryOptions;
-import org.hibernate.search.v6poc.backend.lucene.document.model.impl.RangeQueryOptions;
 import org.hibernate.search.v6poc.backend.lucene.types.impl.formatter.LocalDateFieldFormatter;
 
 public final class LocalDateFieldQueryFactory implements LuceneFieldQueryFactory {
 
-	private final LocalDateFieldFormatter localDateFieldFormatter;
+	private LocalDateFieldFormatter formatter;
```
[the remark this answers is not in the document]
Comment (gsmet) [reply backend-lucene/src/main/java/org/hibernate/search/v6poc/backend/lucene/types/impl/predicate/LocalDateFieldQueryFactory.java]: No reason, a bad copy/paste, fixed.

File: backend-lucene/src/main/java/org/hibernate/search/v6poc/backend/lucene/document/model/impl/LuceneFieldEncoding.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+import java.util.Collections;
+import java.util.Set;
+
+import org.apache.lucene.document.Document;
+import org.hibernate.search.v6poc.backend.lucene.document.impl.LuceneDocumentBuilder;
+
+/**
+ * @author Guillaume Smet
+ */
+public interface LuceneFieldEncoding<T> {
```
[the remark this answers is not in the document]
Comment (gsmet) [reply backend-lucene/src/main/java/org/hibernate/search/v6poc/backend/lucene/document/model/impl/LuceneFieldEncoding.java]: Done.

File: backend-lucene/src/main/java/org/hibernate/search/v6poc/backend/lucene/types/impl/dsl/StringIndexSchemaFieldContext.java [hunk 4]
```diff
@@ -59,16 +59,18 @@ protected void contribute(DeferredInitializationIndexFieldAccessor<String> acces
...
 				getFieldName(),
-				StringFieldFormatter.INSTANCE,
+				formatter,
 				new StringFieldEncoding(
 						sortable,
 						getFieldType( getStore(), analyzer != null ),
 						analyzerOrNormalizer
 				),
-				new StringFieldQueryFactory( analyzerOrNormalizer, analyzer != null, queryBuilder ),
+				new StringFieldQueryFactory( formatter, analyzer != null, queryBuilder ),
```
[the remark this answers is not in the document]
Comment (gsmet) [reply backend-lucene/src/main/java/org/hibernate/search/v6poc/backend/lucene/types/impl/dsl/StringIndexSchemaFieldContext.java]: Yes, I'm a bit worried about the size of the classes in the end. I think we'll have to implement more features to see how it goes.

File: backend-lucene/src/main/java/org/hibernate/search/v6poc/backend/lucene/types/impl/formatter/StringFieldFormatter.java [hunk 5]
```diff
@@ -6,17 +6,32 @@
...
+	public String normalize(String absoluteFieldPath, String value) {
+		if ( value == null ) {
+			return null;
+		}
+
+		if ( analyzerOrNormalizer == null ) {
+			return value;
+		}
+
+		return AnalyzerUtils.analyzeSortableValue( analyzerOrNormalizer, absoluteFieldPath, value );
```
[the remark this answers is not in the document]
Comment (gsmet) [reply backend-lucene/src/main/java/org/hibernate/search/v6poc/backend/lucene/types/impl/formatter/StringFieldFormatter.java]: Changed it to `normalize()` and improved the logging too.

**Comment by gsmet** [comment]
@yrodiere force-pushed an update fixing your comments. I changed `types.impl.*` to `types.*.impl` as I think it's more future-proof in the end. Lots of fun doing so!

**Comment by yrodiere** [comment]
Looks good. Merged, thanks!

**Pull request merged**